packages/@ember/-internals/runtime/lib/mixins/array.d.ts

```diff
@@ -0,0 +1,103 @@
+type Value<T, K extends string> = K extends keyof T ? T[K] : unknown;
+
+export type NatifiedArray<T> = T[] & NativeArray<T>;
+
+interface EmberArray<T> {
+  length: number;
+  objectAt(idx: number): T | undefined;
+}
+declare class EmberArray<T> {
+  objectsAt(indexes: number[]): Array<T | undefined>;
+  get firstObject(): T | undefined;
+  get lastObject(): T | undefined;
+  slice(beginIndex?: number, endIndex?: number): NatifiedArray<T>;
+  indexOf(object: T, startAt?: number): number;
+  lastIndexOf(object: T, startAt?: number): number;
+  forEach<Target>(
+    callback: (this: Target, item: T, index: number, arr: this) => void,
+    target?: Target
+  ): this;
+  getEach<K extends string>(key: K): NatifiedArray<Value<T, K>>;
+  setEach<K extends string>(key: K, value: Value<T, K>): this;
+  map<R, Target>(
+    callback: (this: Target, item: T, index: number, arr: this) => R,
+    target?: Target
+  ): NatifiedArray<R>;
+  mapBy<K extends string>(key: K): NatifiedArray<Value<T, K>>;
+  filter<Target>(
+    callback: (this: Target, item: T, index: number, arr: this) => boolean,
+    target?: Target
+  ): NatifiedArray<T>;
+  reject<Target>(
+    callback: (this: Target, item: T, index: number, arr: this) => boolean,
+    target?: Target
+  ): NatifiedArray<T>;
+  filterBy(key: string): NatifiedArray<T>;
+  rejectBy(key: string): NatifiedArray<T>;
+  find<Target = null>(
+    callback: (this: Target, item: T, index: number, arr: this) => boolean,
+    target?: Target
+  ): T | undefined;
+  findBy<K extends string>(key: K, value?: Value<T, K>): T | undefined;
+  every<Target = null>(
+    callback: (this: Target, item: T, index: number, arr: this) => boolean,
+    target?: Target
+  ): boolean;
+  isEvery<K extends string>(key: K, value?: Value<T, K>): boolean;
+  any<Target = null>(
+    callback: (this: Target, item: T, index: number, arr: this) => boolean,
+    target?: Target
+  ): boolean;
+  isAny<K extends string>(key: K, value?: Value<T, K>): boolean;
+  reduce<V>(
+    callback: (summation: V, current: T, index: number, arr: this) => V,
+    initialValue?: V
+  ): V;
+  invoke<K extends string>(
+    methodName: K,
+    ...args: Value<T, K> extends (...args: any[]) => any ? Parameters<Value<T, K>> : unknown[]
+  ): NatifiedArray<Value<T, K> extends (...args: any[]) => any ? ReturnType<Value<T, K>> : unknown>;
+  toArray(): T[];
+  compact(): NatifiedArray<Exclude<T, null>>;
+  includes(object: T, startAt?: number): boolean;
+  sortBy(key: string): T[];
+  uniq(): NatifiedArray<T>;
+  uniqBy(key: string): NatifiedArray<T>;
+  without(value: T): NatifiedArray<T>;
+}
+
+export default EmberArray;
+
+interface MutableArray<T> extends EmberArray<T> {
```

Comment:
Notably, I believe this is backwards of the relationship in Ember itself (per the docs, and per the current types). We should flip it to match.

Reply:
This is correct, but the situation is confusing.

EmberArray is a Mixin for immutable arrays.
MutableArray is a Mixin that extends EmberArray and adds mutability.
NativeArray is the native Array with MutableArray mixed into it.

If there's a better way to handle all of this, let me know.

Reply:
No, you have the overall right shape of it. I thought the relationship was inverted (because in the types we have on DefinitelyTyped, it is!). Thanks for the clarification. 👍🏼